core/src/main/scala/kafka/server/metadata/ZkMetadataCache.scala

```diff
@@ -55,6 +56,64 @@ trait ZkFinalizedFeatureCache {
   def getFeatureOption: Option[Features]
 }
 
+case class MetadataSnapshot(partitionStates: mutable.AnyRefMap[String, mutable.LongMap[UpdateMetadataPartitionState]],
+                            topicIds: Map[String, Uuid],
+                            controllerId: Option[CachedControllerId],
+                            aliveBrokers: mutable.LongMap[Broker],
+                            aliveNodes: mutable.LongMap[collection.Map[ListenerName, Node]]) {
+  val topicNames: Map[Uuid, String] = topicIds.map { case (topicName, topicId) => (topicId, topicName) }
+}
+
+object ZkMetadataCache {
+  /**
+   * Create topic deletions (leader=-2) for topics that are missing in a FULL UpdateMetadataRequest coming from a
+   * KRaft controller during a ZK migration.
+   */
+  def maybeInjectDeletedPartitions(
```

Comment:
perhaps `maybeInjectDeletedPartitionsFromFullMetadataRequest`? Or is that too long?